### ISSUE TITLE
Gradient calculation of tf.reduce_prod runs in CPU, inducing performance impact

### ISSUE THREAD

**Issue by carusyte** [tensorflow/tensorflow#40748]
<em>Please make sure that this is an issue related to performance of TensorFlow.
As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:performance_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): _Yes_
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): _Ubuntu 19.04_
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: _No_
- TensorFlow installed from (source or binary): _binary_
- TensorFlow version (use command below): _2.2.0_
- Python version: _3.7.7_
- Bazel version (if compiling from source): _N/A_
- GCC/Compiler version (if compiling from source): _N/A_
- CUDA/cuDNN version: _CUDA Version 10.2.89_
- GPU model and memory: _GeForce RTX 2080 8G_

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
Gradient of tf.reduce_prod() runs in CPU instead of GPU, which induces lots of memcopy between GPU & CPU as shown in profiler, and slows down entire training / prediction process as GPU & CPU ops interleaves (Kernel Launch Time would dominate in the profiler report).
Pertinent code can be found [here](https://github.com/tensorflow/tensorflow/blob/4f341bb742718721563ce6dccb965c85a1fbdcf5/tensorflow/python/ops/math_grad.py#L270)

**Describe the expected behavior**
Expect the gradient to be calculated in GPU and streamline the whole process (especially in a typical RNN while loop)

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.
_To be provided_

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.
_To be provided_

**Comment by Saduf2019** [comment]
@carusyte 
I ran the code shared by you and face a different error, please find the [gist here](https://colab.research.google.com/gist/Saduf2019/a255d16383bbaa8db62d3f1458556b47/untitled243.ipynb).

**Comment by carusyte** [comment]
@Saduf2019

**Comment by carusyte** [comment]
@Saduf2019 
Thanks for your prompt reply. The link to the code I posted is just an excerpt of the tensorflow 2.2.0 codebase, which I meant to highlight the block starting with `with ops.device("/cpu:0")` under `_ProdGrad(op, grad)` function.
```
def _ProdGrad(op, grad):
  ...omitted code for brevity...

  # Pack all reduced dimensions into a single one, so we can perform the
  # cumprod ops. If the reduction dims list is empty, it defaults to float32,
  # so we need to cast here.  We put all the shape-related ops on CPU to avoid
  # copying back and forth, and since listdiff is CPU only.
  with ops.device("/cpu:0"):
    rank = array_ops.rank(op.inputs[0])
    reduction_indices = (reduction_indices + rank) % rank
    reduced = math_ops.cast(reduction_indices, dtypes.int32)
    idx = math_ops.range(0, rank)
    other, _ = array_ops.setdiff1d(idx, reduced)
    perm = array_ops.concat([reduced, other], 0)
    reduced_num = math_ops.reduce_prod(array_ops.gather(input_shape, reduced))
    other_num = math_ops.reduce_prod(array_ops.gather(input_shape, other))

  ...omitted code for brevity...
```

I thought this could be the cause of the performance bottleneck, resulting in high percentage of Kernel Launch time and memcpy operations, breaking the streamlined computation in GPU, as shown in the following screenshots from the profile:

![image](https://user-images.githubusercontent.com/1680881/85830228-d39a5e80-b7be-11ea-86a4-45a633cb20ba.png)

![image](https://user-images.githubusercontent.com/1680881/85830600-7521b000-b7bf-11ea-80d2-2c85cd8086f5.png)

_In the trace view above, GPU operation stripes are shaped like a comb, which makes me suspect some kind of op must have been delegated to the CPU, interrupting GPU processing. Going further, I found several `Prod_grad` related ops under the CPU section, amongst other less recognizable ops_ 

![image](https://user-images.githubusercontent.com/1680881/85831574-13fadc00-b7c1-11ea-92a8-d19313746248.png)

However, as I tried to dig a little deeper / get around this issue, by replacing `tf.reduce_prod` with a combination of `tf.math.cumprod`, `tf.slice` etc., the above `Prod_grad` ops went away under the CPU section, but the overall avg. 788 ms step time and its composition didn't make a difference.

Therefore, I assumed `tf.reduce_prod` might not be the culprit for now... That being said, I'm kind of lost as to where to look for the contributor to the high Kernel Launch time. Here's the [profile](https://github.com/tensorflow/tensorflow/files/4835982/profile.zip) for tensorboard if you would like to investigate further. Note that this is just a simplified model where each step takes 788 ms. In my targeted deeper model it takes more than 2 seconds per step which is greatly hindered by the Kernel Launch time.

**Comment by sushreebarsa** [comment]
@carusyte Could you please try on the latest stable TensorFlow version 2.7.0 and let us know if it helps?Thanks!
